app/controllers/glossary_terms_controller.rb

```diff
@@ -0,0 +1,150 @@
+    @objects = @glossary_term.images
+  end
+
+  # ---------- Actions to Display forms -- (new, edit, etc.) -------------------
+
+  def new
+    @copyright_holder = @user.name
+    @copyright_year = Time.now.utc.year
+    @upload_license_id = @user.license_id
+    @licenses = License.current_names_and_ids(@user.license)
+  end
+
+  def edit
+    @glossary_term = GlossaryTerm.find(params[:id].to_s)
+  end
+
+  # ---------- Actions to Modify data: (create, update, destroy, etc.) ---------
+
+  def create
+    glossary_term = \
+      GlossaryTerm.new(user: @user, name: params[:glossary_term][:name],
+                       description: params[:glossary_term][:description])
+
+    if params[:glossary_term][:upload_image]
+      glossary_term.add_image(process_image(image_args))
+    end
+    glossary_term.save
+    redirect_to(glossary_term_path(glossary_term.id))
+  end
+
+  def update
+    glossary_term = GlossaryTerm.find(params[:id].to_s)
+    glossary_term.attributes = params[:glossary_term].
+                               permit(:name, :description)
+    glossary_term.user = @user
+    glossary_term.save
+    redirect_to(glossary_term_path(glossary_term.id))
+  end
+
+  def destroy
+    return unless (@glossary_term = GlossaryTerm.find(params[:id]))
+
+    if in_admin_mode?
+      @glossary_term.destroy
+      flash_notice(:runtime_destroyed_id.t(type: GlossaryTerm,
+                                           value: params[:id]))
+    else
+      flash_warning(:permission_denied.t)
+    end
+
+    redirect_to(glossary_terms_path)
+  end
+
+  # ---------- Non-standard REST Actions ---------------------------------------
+
+  # Show past version of GlossaryTerm.
+  # Accessible only from show_glossary_term page.
+  def show_past
+    if @glossary_term = find_or_goto_index(GlossaryTerm, params[:id].to_s)
+      if params[:version]
+        @glossary_term.revert_to(params[:version].to_i)
+      else
+        flash_error(:show_past_location_no_version.t)
+        redirect_to(glossary_term_path(@glossary_term.id))
+      end
+    end
+  end
+
+  # ---------- Public methods (unrouted) ---------------------------------------
+
+  ##############################################################################
+
+  private
+
+  # --------- Filters
+
+  # --------- Other private methods
+
+  def image_args
+    Rails.env.test? ? permit_upload_image_param : strong_upload_image_param
+  end
+
+  def strong_upload_image_param
+    args = {
+      copyright_holder: params[:copyright_holder],
+      when: Time.local(params[:date][:copyright_year]).utc,
+      license: License.safe_find(params[:upload][:license_id]),
+      user: @user,
+      image: params[:glossary_term][:upload_image]
+    }
+  end
+
+  def permit_upload_image_param
+    args = strong_upload_image_param
+    # Remove "permitted: false" from this param so that model can mass assign it
+    args[:image] = params[:glossary_term][:upload_image][:image]
+    args
+  end
```

Comment:
Allows request (functional) testing whether an image is actually uploaded and associated with a GlossaryTerm. See `GlossaryTermsControllerTest#test_create_upload_image`. 
**Is there a better way to do this?**
I think that without this (or something similar), the model cannot mass assign `[:glossary_term][:upload_image][:image]` because Rails marks that param as `permitted: false`(if it doesn't come from a form).
The only alternative I can think of is a system test using Capybara. That's slow, a little painful to code, and @mo-nathan wanted to avoid general use of Capybara.

Reply:
I agree this smells wrong.  We're supposed to be avoiding the explicit reference to Rails.env.test? in the code these days.  But I'm too exhausted to understand the issue fully so I can't suggest alternatives.  I remember feeling like I had no choice but the jury-rig parameters for testing image uploads in the past, too.  So maybe I'm the wrong person to ask!

Reply:
Another problem with this approach is that we could not test what would happen if a user put bad stuff into a form. (But I think it's unlikely we'd want to test that.)
But it's not a terribly bad smell and I dislike slowing the test suite by adding more Capybara tests. The most recent full test run on my machine took 7:34, which is about 2x what it took when I first started working on MO.
I think the options are:
- a slightly smelly functional test
- a system (Capybara) test
- no test
I'll do whatever other folks want.